File: 3702-longest-subsequence-with-non-zero-bitwise-xor/3702-longest-subsequence-with-non-zero-bitwise-xor.py
class Solution:
    def longestSubsequence(self, nums: List[int]) -> int:
        total_xor = 0
        all_zero = True
        for num in nums:
            total_xor ^= num
            if num > 0:
                all_zero = False
        
        if total_xor:
            return len(nums)
        
        if all_zero:
            return 0
        
        return len(nums) - 1

# array, bit manipulation
# time: O(n)
# space: O(1)

File: 3702-longest-subsequence-with-non-zero-bitwise-xor/3702-longest-subsequence-with-non-zero-bitwise-xor.go
func longestSubsequence(nums []int) int {
    totalXor := 0
    allZero := true
    for _, num := range nums {
        totalXor ^= num
        if num > 0 {
            allZero = false
        }
    }

    if totalXor > 0 {
        return len(nums)
    }

    if allZero {
        return 0
    }

    return len(nums) - 1
}

// array, bit manipulation
// time: O(n)
// space: O(1)

// n.b
// so, here the pattern
// first, know some xor properties: a ^ 0 = a, a ^ a = 0, a ^ b = b ^ a, a ^ (b ^ c) = (a ^ b) ^ c
// if the total xor is not 0, basically the longest subseq is the nums's length
// here's the magic
// if the total xor is 0, if there exists a minimum ONE num, with num > 0, the max length is the num's length - 1
// else, the max length is 0
// it's hard to believe at first
// try to simulate it on paper, and then the pattern makes sense